### ISSUE TITLE
Nix install instructions are incorrect'ish

### ISSUE THREAD

**Issue by peti** [simonmichael/hledger#980]
https://hledger.org/download.html says that hledger ought to be installed with Nix by running

    $ nix-env -iA nixpkgs.haskellPackages.hledger nixpkgs.haskellPackages.hledger-ui \
                  nixpkgs.haskellPackages.hledger-web

That command is going to work most of the time, but sometimes it's not because not all installations have Nixpkgs under a `nixpkgs` prefix. Some use `nixos` instead. The proper way to specify that you're using a Nixpkgs package is `nix-env -f "<nixpkgs>"`. Then these prefixes can be omitted.

Furthermore, hledger has top-level attributes that reference way fewer dependencies, i.e. these don't depend on `ghc` at run-time. So I'd recommend using:

    $ nix-env -f "<nixpkgs>" -iA hledger hledger-ui hledger-web

**Comment by peti** [comment]
P.S.: The command

    $ nix-env -i hledger hledger-ui hledger-web

is even simpler and achieves the same result. It's slower and less idiomatic, though. For our purposes it might actually be preferable, though.

**Comment by simonmichael** [comment]
Thanks! The last one looks great. How much slower ?

Related to nix, I wonder if any mac users know if https://github.com/simonmichael/hledger/issues/613 is resolved.

**Comment by peti** [comment]
> Thanks! The last one looks great. How much slower ?

Depends on your hard drive / SSD / state of the cache. `nix-env -i` might run several seconds before it's ready to start installing anything. `nix-env -iA` does that almost instanteuously. 

> Related to nix, I wonder if any mac users know if #613 is resolved.

According to [Hydra](https://hydra.nixos.org/job/nixpkgs/trunk/haskellPackages.hledger-ui.x86_64-darwin), `hledger-ui` compiles fine on Darwin.

**Comment by simonmichael** [comment]
Thanks. I've updated the download page with the simpler command.

**Comment by simonmichael** [comment]
It looked so simple I was inspired to try it. I didn't find nix in homebrew, but guessed that nix-env could be obtained from https://nixos.org/nix/, which worked. And now I can do:
```
$ nix-env -i hledger hledger-ui hledger-web 
installing 'hledger-1.13.2'
installing 'hledger-ui-1.13.1'
installing 'hledger-web-1.13'
...
```
I guess this will change to 1.14 before too long.

**Comment by simonmichael** [comment]
(And that makes sense, because the "nixpkgs unstable package" badge is still showing 1.13.2.)

**Comment by peti** [comment]
The 1.4 versions are building already: https://hydra.nixos.org/eval/1507546?filter=hledger&compare=1507440. The packages will show up in the `nixpkgs-unstable` channel within the next few hours (assuming those builds work properly).

**Comment by simonmichael** [comment]
End result: working hledger binaries on mac in less than a minute with only 319M used, in /nix. Very nice!

**Comment by simonmichael** [comment]
If I'm looking in the right place, I believe the 1.14 binaries built yesterday, but the command still seems to see only 1.13. Is there something we should add to force a refresh ?

**Comment by peti** [comment]
> The command still seems to see only 1.13. Is there something we should add to force a refresh?

You'll need to run `nix-channel --update`. If you are on `nixpkgs-unstable`, then you'll have the new binaries already. https://howoldis.herokuapp.com/ has the current state of the various NixOS channels.

**Comment by simonmichael** [comment]
Thanks, I've added this to the page.

**Comment by simonmichael** [comment]
@peti, sorry for the stream of questions but the nix badges on the hledger download page, which we get from https://repology.org/project/hledger/badges seem not to be updating. Would you have any idea why ?

**Comment by peti** [comment]
Unfortunately, I don't know how that works. I trust it's going to be updated eventually, but I'm not sure how often that happens or what triggers it.

**Comment by simonmichael** [comment]
Ok, thanks!

**Comment by peti** [comment]
I see that the badge is now up-to-date. :-)

**Comment by simonmichael** [comment]
Now we just need them to be green/red like the rest.

**Comment by peti** [comment]
I've noticed that the download instructions for Nix now say: 

    nix-channel --update && nix-env -i hledger hledger-ui hledger-web

I don't think that's a good idea. Many people don't follow any particular channel but rather check out Nixpkgs from Github, so running `nix-channel` makes no sense for them. Furthermore, many people follow a channel that will not receive any updates of `hledger` this way, like the stable release channels, for example. Last but not least, it's certainly inconsistent that the Nix command includes a kind-of "update global package manager data" command in its instructions, but the instructions for ArchLinux, Debian, or Ubuntu do not, i.e. there is no `apt update && ...`   anywhere in the page.

**Comment by simonmichael** [comment]
Thanks. There might be some inconsistency indeed. The goal on this page is to have a command that as reliably as possible installs the best hledger version available by that method, ie the hledger version advertised just to the right. In practical terms, my experience so far is that with cabal and nix-env, eg, I normally have to run an update command to get the advertised version. With some others, that's automatic (stack, docker) or semiautomatic (brew, which prints a helpful message). With linux distros, significant updates to the hledger packages are very infrequent, also it's likely that the user has done an update somewhat recently as part of system maintenance. 

Obviously I'm a bit clueless about some of these platforms (which makes me a useful test subject). What are the potential bad consequences of running nix-channel --update ?

**Comment by simonmichael** [comment]
And more to the point, what's a better, still reasonably short, command that will get the best available version ?

**Comment by peti** [comment]
I would not list any command other than `nix-env -i hledger ...`. User's of Nix will know how to keep their package database up-to-date, and if they don't know then there's ample documentation available on the Nix website that explains them in great detail how to do it. It doesn't feel like a good idea to me to try and duplicate that information.

> What are the potential bad consequences of running nix-channel --update?

I don't think that anything really bad can happen if you run that command. You may end up downloading and installing a new version of Nixpkgs, though, that the following `nix-env -i ...` command actually doesn't use, which would be a waste of resources. Arguably, people who have a setup like that will typically recognize on their own that the channel update is not useful to them and just won't run it. Still, say that running `nix-channel --update` is a necessary (or even recommended) step to take before installing `hledger`. It's inconsistent with the documentation relating to the other package managers and it's also somewhat technically inaccurate.

**Comment by simonmichael** [comment]
> On Mar 10, 2019, at 3:39 AM, Peter Simons <notifications@github.com> wrote
> I would not list any command other than nix-env -i hledger .... User's of Nix will know how to keep their package database up-to-date, and if they don't know then there's ample documentation available on the Nix website that explains them in great detail how to do it. It doesn't feel like a good idea to me to try and duplicate that information.
> 
Coming from a Nix guru, I take this seriously. Some additional considerations:

The download page aims to smooth the path for folks new to hledger, and probably new to Nix as well. Every bit of friction loses potential users/contributors. People who know their packaging system well, mainly just need the package names; they will know what to do. 

I, motivated, and a Nix newbie, tried the "nix-env -i hledger hledger-ui hledger-web" command, several times over two days, but I kept getting the old release of hledger even though the page said I would get the new one, and when I searched Nix packages online I could see it had built the new one. That's how it works, you could say. But you can see how this will be unsatisfying to a newcomer. 

So I need to do something better that avoids this. Any ideas ? I could add help text, but it doesn't seem better:

nix-env -i hledger hledger-ui hledger-web   (unstable users, please run nix-channel --update to get the latest)

> What are the potential bad consequences of running nix-channel --update?
> 
> I don't think that anything really bad can happen if you run that command. You may end up downloading and installing a new version of Nixpkgs, though, that the following nix-env -i ... command actually doesn't use, which would be a waste of resources. Arguably, people who have a setup like that will typically recognize on their own that the channel update is not useful to them and just won't run it. Still, say that running nix-channel --update is a necessary (or even recommended) step to take before installing hledger. It's inconsistent with the documentation relating to the other package managers and it's also somewhat technically inaccurate.
> 

(I'm assuming you meant "Still, don't say" above).

**Comment by simonmichael** [comment]
Would specifying the versions be enough, as with cabal-install ? Eg

`nix-env -i hledger-1.14.1 hledger-ui-1.14 hledger-web-1.14`

It seems to work right now, but I'm guessing it might fail if I hadn't done the channel update first.

**Comment by simonmichael** [comment]
Wanting to educate myself about Nix, I spent some time reading docs and chatting in #nixos today. I thought we had a winner:

    nix-env -f channel:nixos-unstable -iA hledger hledger-ui hledger-web

But testing, I found that this and the previous variants were doing an awful lot of building from source and taking up 3.6G unpacked, unlike [before](https://github.com/simonmichael/hledger/issues/980#issuecomment-468939754). It seems the nix [darwin] hledger builds broke right after that due to a [blaze-builder issue](https://hydra.nixos.org/build/90567978). So a transient breakage like this in unstable, which I imagine is pretty common, can transform a normally mild-mannered install command like the above into a disk-filling build monster, which would be a poor experience.

Two solutions were mentioned: 

- use a pinned known-good version, eg the first successful build after a hledger release. Once installed, users would have to take some special steps to update it:

       nix-env -f https://github.com/NixOS/nixpkgs/archive/GITREV.tar.gz -iA hledger hledger-ui hledger-web

- add a default.nix to the project, upload our own official builds to cachix.org, and use:

       nix-env -f https://github.com/simonmichael/hledger/archive/master.tar.gz -iA hledger hledger-ui hledger-web

Otherwise, I think we'd need some advice about this issue on the download page. Eg recommend trying --dry-run first, and if the size is abnormally large wait for build to be fixed.. a link or badge showing accurate unstable build status.. etc.

**Comment by simonmichael** [comment]
Today I've made these changes at https://hledger.org/download :
- removed nix-channel --update as you suggested
- added package version numbers. I am guessing these will cause it to fail until channel is updated, which will be an improvement. The command implies use of nixpkgs-unstable now.
- added a note about nix-channel --update and the large downloads due to build failures issue.

**Comment by Infinisil** [comment]
It's not a large download issue: Even with it being a large download, the build will just end up failing. The reason is that Nix unfortunately doesn't have any notion of cached failure, only successful builds are cached. Therefore I suggest rewording it to something like "Builds should be cached, unless in case of CI issues, so check with --dry-run first".

**Comment by simonmichael** [comment]
Thanks!

**Comment by peti** [comment]
Very cool, the new install instructions for Nix seem pretty good. :-)

**Comment by simonmichael** [comment]
Today I tried the pinned option. Since there has been one successful build of the desired hledger release, this works well, adding just 200M to /nix:

    nix-env -iA hledger hledger-ui hledger-web -f https://github.com/NixOS/nixpkgs/archive/9c74e2.tar.gz

I'm not sure if there's any significant downside, so I'm tempted to show only this command.

**Comment by simonmichael** [comment]
Our nix one-line install command hasn't been tested recently but it still works as well as it did AFAIK.
